src/jdk.httpserver/share/classes/sun/net/httpserver/ServerImpl.java

```diff
@@ -961,9 +963,24 @@ void requestStarted (HttpConnection c) {
     }
 
     void markIdle(HttpConnection c) {
-        c.idleStartTime = System.currentTimeMillis();
-        c.setState(State.IDLE);
-        idleConnections.add(c);
+        Boolean close = false;
+
+        idleConnectionLock.lock();
+        if (idleConnections.size() >= MAX_IDLE_CONNECTIONS) {
```

Comment:
Just a general point here is that you probably should use try-finally and do the unlock in the finally block, just in case something between lock and unlock throws or the control flow is changed for some reason, like an early return.